### PR TITLE
Several notification bugfixes and tweaks

### PR DESCRIPTION
**Pull request by andrewconner**

### CODE REVIEW

File: kifi-backend/modules/shoebox/public/css/main.css [hunk 1]
```diff
@@ -16,7 +16,7 @@ a[href] {color: #0089d5; text-decoration: none}
 
 html {height: 100%; text-align: center; overflow-y: hidden; overflow-x: auto}
 body {height: 100%; min-width: 1024px; margin: 0 auto; padding: 0; background: #fafafa; font-size: 0; text-align: left; position: relative}
-.page-cols {display: table; table-layout: fixed; min-width: 100%; height: 100%; overflow: hidden}
+.page-cols {display: table; table-layout: fixed; min-width: 100%; height: 100%; overflow: hidden; -webkit-font-smoothing: subpixel-antialiased;}
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/public/css/main.css]: Sneaking this in... this fixes a bug in Safari on kifi.com where fonts lose aliasing during animations.

File: packrat/scripts/slider2.js [hunk 2]
```diff
@@ -728,7 +728,7 @@ slider2 = function() {
     counts: function(o) {
       if (!$slider) return;
       var $btns = $slider.find(".kifi-slider2-dock-btn");
-      [[".kifi-slider2-notices", Math.max(0, o.n - o.m)],
+      [[".kifi-slider2-notices", Math.max(0, o.n - (o.m > 0 ? 1 : 0))],
```
Comment (andrewconner) [inline packrat/scripts/slider2.js]: This is still not perfect, but is closer to expected behavior. `m` is the number of unread messages inside of threads on this page. Previously, when there were several messages, `n - m` can be less than 0, so the count goes away.

File: kifi-backend/modules/shoebox/public/css/main.css [hunk 3]
```diff
@@ -16,7 +16,7 @@ a[href] {color: #0089d5; text-decoration: none}
 
 html {height: 100%; text-align: center; overflow-y: hidden; overflow-x: auto}
 body {height: 100%; min-width: 1024px; margin: 0 auto; padding: 0; background: #fafafa; font-size: 0; text-align: left; position: relative}
-.page-cols {display: table; table-layout: fixed; min-width: 100%; height: 100%; overflow: hidden}
+.page-cols {display: table; table-layout: fixed; min-width: 100%; height: 100%; overflow: hidden;}
```
Comment (2is10) [inline kifi-backend/modules/shoebox/public/css/main.css:19]: Thanks for adding a semicolon at the end of the declaration;.

File: packrat/scripts/notifier.js [hunk 4]
```diff
@@ -6,14 +6,18 @@
 // @require scripts/render.js
 
 var notifier = {
+  removeByUnqiueId: function(uniqueId) {
+    KifiNotification.removeByUnqiueId(uniqueId);
```
Comment (2is10) [inline packrat/scripts/notifier.js]: spelling

File: packrat/html/notify_box.html [hunk 5]
```diff
@@ -1,5 +1,5 @@
 
-<div class="kifi-notify-item-wrapper {{popupClass}}" style="display:none">
+<div class="kifi-notify-item-wrapper {{popupClass}}" style="display:none" data-uniqueid="{{uniqueId}}">
```
Comment (2is10) [inline packrat/html/notify_box.html]: I incorrectly guessed the meaning of “unique” here. At first I thought you were giving each notification DOM element a unique ID for some reason. How about `associatedId` or `relatedId`?

Comment (andrewconner) [inline packrat/html/notify_box.html]: That works. unique per notification

Comment (2is10) [inline packrat/html/notify_box.html]: Also, the custom for `data-` attributes is to use a hyphen between words, like `data-unique-id`, and then use camel case when accessing them in script.

File: packrat/main.js [hunk 6]
```diff
@@ -493,25 +495,27 @@ api.port.on({
...
     });
   },
-  set_message_read: function(o, _, tab) {
-    var d = pageData[tab.nUri];
-    var unreadNotification = false;
-    for (var i = 0; i < notifications.length; i++) {
-      if (notifications[i].unread && (notifications[i].thread == o.threadId || notifications[i].id == o.messageId)) {
-        unreadNotification = true;
-        o.messageId = notifications[i].id;
+  set_message_rendered: function(o, _, tab) {
```
Comment (2is10) [inline packrat/main.js]: How about just `message_rendered` since we don’t actually store the fact that it was rendered anywhere.

File: packrat/scripts/notifier.js [hunk 7]
```diff
@@ -6,14 +6,18 @@
...
   show: function(data) {
     var o = data;
     switch (data.category) {
       case "message":
+        KifiNotification.removeByUnqiueId(o.thread, {fade: false});
         KifiNotification.add({
           title: o.author.firstName + " " + o.author.lastName,
           subtitle: "Sent you a new Kifi Message",
-          contentHtml: o.text,
+          contentHtml: o.text.trim().length > 200 ? o.text.substring(0, 200).trim() + "…" : o.text,
```
Comment (2is10) [inline packrat/scripts/notifier.js]: It’s a good idea to use two different numbers here, so that you avoid replacing just a single character (such as a period) with an ellipsis. I’d recommend a spread of anywhere from 3 to 20 characters.

Comment (2is10) [inline packrat/scripts/notifier.js]: Don’t we trim content on its way in? Not sure we need to do it again before display.

File: packrat/scripts/notifier.js [hunk 8]
```diff
@@ -127,7 +135,9 @@ var KifiNotification = {
       fadeOutMs = params.fadeOutMs || 300,
       manualClose = unbindEvents;
 
-    $item.data("beforeClose")($item, manualClose);
+    if ($item.data("beforeClose")) {
+      $item.data("beforeClose")($item, manualClose);
+    }
```
Comment (2is10) [inline packrat/scripts/notifier.js]: shorter to write `($item.data("beforeClose") || $.noop)(...);`

File: packrat/scripts/slider2.js (hunk 2 above)
Comment (2is10) [inline packrat/scripts/slider2.js]: `m` should be the number of threads with unread messages. it used to be. if it’s not, then I think we should change it back.

File: packrat/main.js [hunk 9]
```diff
@@ -643,6 +650,12 @@ api.port.on({
   }
 });
 
+function removeNotificationPopup(uniqueId) {
```
Comment (2is10) [inline packrat/main.js]: `removeNotificationPopups` with an “s”?

File: packrat/main.js [hunk 10]
```diff
@@ -1067,10 +1080,23 @@ function clone(o) {
   return c;
 }
 
+function whenTabFocused(tab, idx, callback) {
```
Comment (2is10) [inline packrat/main.js]: `idx` appears to be unused

Comment (andrewconner) [inline packrat/main.js]: Whoops, `tab.id` below is meant to be `idx`. Fixing.

Comment (2is10) [inline packrat/main.js]: Okay if we call it `key` instead of `idx`? `idx` makes me think of an array index. Here and at line 1095. Thanks.

File: packrat/scripts/notifier.js [hunk 11]
```diff
@@ -140,16 +150,26 @@ var KifiNotification = {
...
   },
 
   removeSpecific: function($item, params, unbindEvents) {
+    api.port.emit("remove_notification", {uniqueId: $item.data("uniqueid")});
     KifiNotification.fadeItem($item, params || {}, unbindEvents);
   },
 
+  removeByUniqueId: function(uniqueId, params) {
+    var $wrap = $("#kifi-notify-notice-wrapper");
+    $wrap.find(".kifi-notify-item-wrapper[data-uniqueId='" + uniqueId + "']").each(function(i,e) {
```
Comment (2is10) [inline packrat/scripts/notifier.js]: Don’t need to list and name the anonymous function parameters if you don’t use them.

File: packrat/scripts/notifier_scout.js [hunk 12]
```diff
@@ -11,6 +11,13 @@ var notifierScout = notifierScout || function() {  // idempotent for Chrome
         }
       }
     },
+    remove_notification: function(uniqueId) {
+      if (document.querySelectorAll(".kifi-notify-item-wrapper[data-uniqueId='" + uniqueId + "']").length > 0) {
```
Comment (2is10) [inline packrat/scripts/notifier_scout.js]: shorter:
`if (document.querySelector(".kifi-notify-item-wrapper[data-uniqueId='" + uniqueId + "']")) {`

File: packrat/html/notify_box.html (hunk 5 above)
Comment (2is10) [inline packrat/html/notify_box.html]: The custom I’m referring to originates in the [HTML 5 spec](http://www.w3.org/TR/2011/WD-html5-20110525/elements.html#embedding-custom-non-visible-data-with-the-data-attributes). To be clear, use hyphens between words in the HTML and whenever you are referring specifically to the attribute name (e.g. in an attribute selector in CSS or JS, a `.getAttribute` DOM call, an `.attr` jQuery call) and use camel case when getting or setting the data value via the `.dataset` DOM property or a `.data` jQuery call.

File: packrat/scripts/slider2.js (hunk 2 above)
Comment (2is10) [inline packrat/scripts/slider2.js]: How about reverting this line in this pull request and addressing it separately? Seems unrelated to the rest.

**Comment by 2is10** [comment]
I’d be happy to take another look at this later, if you’d like. Just let me know.

File: packrat/scripts/notifier_scout.js [hunk 13]
```diff
@@ -11,6 +11,13 @@ var notifierScout = notifierScout || function() {  // idempotent for Chrome
         }
       }
     },
+    remove_notification: function(associatedId) {
+      if (document.querySelectorAll(".kifi-notify-item-wrapper[data-associated-id='" + associatedId + "']")) {
```
Comment (2is10) [inline packrat/scripts/notifier_scout.js:15]: `.querySelector` will return null (what you want). `.querySelectorAll` will return an empty array.
